components/layout/flow/inline/construct.rs

```diff
@@ -233,27 +242,67 @@ impl InlineFormattingContextBuilder {
 
     pub(crate) fn start_inline_box(
         &mut self,
-        inline_box: InlineBox,
+        inline_box_creator: impl FnOnce() -> ArcRefCell<InlineBox>,
+        block_in_inline_splits: Option<Vec<ArcRefCell<InlineItem>>>,
+        old_layout_box: Option<LayoutBox>,
+    ) {
+        // If there is an existing undamaged layout box that's compatible, use the `InlineBox` within it.
+        if let Some(LayoutBox::InlineLevel(inline_level_box)) = old_layout_box {
+            let old_block_in_inline_splits: Vec<ArcRefCell<InlineBox>> = inline_level_box
+                .iter()
+                .rev() // reverse to facilate the `Vec::pop` operation
+                .filter_map(|inline_item| match &*inline_item.borrow() {
+                    InlineItem::StartInlineBox(inline_box) => Some(inline_box.clone()),
+                    _ => None,
+                })
+                .collect();
+
+            debug_assert!(
+                old_block_in_inline_splits.is_empty() ||
+                    old_block_in_inline_splits.len() == inline_level_box.len(),
+                "Create inline box with incompatible `old_layout_box`"
+            );
+
+            self.start_inline_box_internal(
+                inline_box_creator,
+                block_in_inline_splits,
+                old_block_in_inline_splits,
+            );
+        } else {
+            self.start_inline_box_internal(inline_box_creator, block_in_inline_splits, vec![]);
+        }
+    }
+
+    pub fn start_inline_box_internal(
+        &mut self,
+        inline_box_creator: impl FnOnce() -> ArcRefCell<InlineBox>,
         block_in_inline_splits: Option<Vec<ArcRefCell<InlineItem>>>,
+        mut old_block_in_inline_splits: Vec<ArcRefCell<InlineBox>>,
     ) {
-        self.push_control_character_string(inline_box.base.style.bidi_control_chars().0);
+        let inline_box = old_block_in_inline_splits
+            .pop()
+            .unwrap_or_else(inline_box_creator);
+        self.push_control_character_string(inline_box.borrow().base.style.bidi_control_chars().0);
 
         // Don't push a `SharedInlineStyles` if we are pushing this box when splitting
         // an IFC for a block-in-inline split. Shared styles are pushed as part of setting
         // up the second split of the IFC.
-        if inline_box.is_first_split {
+        if inline_box.borrow().is_first_split {
             self.shared_inline_styles_stack
-                .push(inline_box.shared_inline_styles.clone());
+                .push(inline_box.borrow().shared_inline_styles.clone());
         }
```

Comment:
Better just borrow once?


```suggestion
        let borrowed_inline_box = inline_box.borrow();
        self.push_control_character_string(borrowed_inline_box.base.style.bidi_control_chars().0);

        // Don't push a `SharedInlineStyles` if we are pushing this box when splitting
        // an IFC for a block-in-inline split. Shared styles are pushed as part of setting
        // up the second split of the IFC.
        if borrowed_inline_box.is_first_split {
            self.shared_inline_styles_stack
                .push(borrowed_inline_box.shared_inline_styles.clone());
        }
        std::mem::drop(borrowed_inline_box);
```